packages/store-ui/src/molecules/NumberInput/NumberInput.tsx

```diff
@@ -0,0 +1,160 @@
+}
+
+const ActionButton = ({
+  children,
+  actionOption,
+  disabled,
+  ariaLabel,
+  onClick,
+}: PropsWithChildren<ActionButtonProps>) => (
+  <Button
+    type="button"
+    data-store-button={actionOption}
+    onClick={(event) => onClick(parseInt(event.currentTarget.value, 10))}
+    disabled={disabled}
+    aria-label={ariaLabel}
+    aria-disabled={disabled}
+  >
+    {children}
+  </Button>
+)
+
+export interface NumberInputProps {
+  /**
+   * The initial value of the input. Should be less than max and greater then min.
+   */
+  defaultValue?: number
+  /**
+   * The minimum value of the input. When the input value reaches the minimum, the decrease's button is disabled.
+   */
+  min?: number
+  /**
+   * The maximum value of the input. When the input value reaches the maximum, the increase's button is disabled.
+   */
+  max?: number
+  /**
+   * Define if the component is disabled.
+   */
+  disabled?: boolean
+  /**
+   * Callback that fires when the current value changes, either by directly changing the input or by buttons' action.
+   */
+  onChange?: (value: number) => void
+  /**
+   * ID to find this component in testing tools (e.g.: cypress, testing library, and jest).
+   *
+   * @default 'store-number-input'
+   */
+  testId?: string
+}
+
+const NumberInput = ({
+  defaultValue,
+  min = -Infinity,
+  max = Infinity,
+  disabled = false,
+  testId = 'store-number-input',
+  onChange,
+}: NumberInputProps) => {
+  const initialValue =
+    defaultValue && defaultValue >= min && defaultValue <= max
+      ? defaultValue
+      : min !== -Infinity
+      ? min
+      : 0
+
+  const [currentValue, setCurrentValue] = useState(initialValue)
+  const [buttonDisabled, setButtonDisabled] = useState<{
+    dec: boolean
+    inc: boolean
+  }>()
+
+  const handleCurrentValueUpdate = useCallback(
+    (newValue: number) => {
+      setCurrentValue(newValue)
+      onChange?.(newValue)
+    },
+    [onChange]
+  )
+
+  const handleInputChange = useCallback(
+    (event: React.FormEvent<HTMLInputElement>) => {
+      if (disabled) return
+
+      const inputValue = event.currentTarget.value
+
+      if (inputValue === '') handleCurrentValueUpdate(initialValue)
+      else handleCurrentValueUpdate(parseInt(inputValue, 10))
+    },
+    [disabled, initialValue, handleCurrentValueUpdate]
+  )
+
+  const handleActionButtonClick = useCallback(
+    (actionOption: 'dec' | 'inc') => {
+      if (currentValue < min) handleCurrentValueUpdate(min)
+      else if (currentValue > max) handleCurrentValueUpdate(max)
+      else {
+        const newValue =
+          actionOption === 'dec' ? currentValue - 1 : currentValue + 1
+
+        handleCurrentValueUpdate(newValue)
+      }
+    },
+    [currentValue, max, min, handleCurrentValueUpdate]
+  )
+
+  useEffect(() => {
+    setButtonDisabled({
+      dec: currentValue <= min || disabled,
+      inc: currentValue >= max || disabled,
+    })
+  }, [currentValue, disabled, min, max])
```

Comment:
I don't think any of this should be on the molecule itself. We want to manage as little state as possible on this library and leave purely the UI to behave as the parent component declares.

What I thought for us to do is to expose different callbacks for users to do the same thing that's here or something different, but on the store side.

The advantage of that is that they might want to, for example, show a dollar sign before the number on the input box, or show the unity (like g, kg, L, etc) after the item. They wouldn't be allowed to do this using this implementation. Also, a common use case is to multiply the value for a unit multiplier.

Consider the following example to understand the requirement: each unit of a wire a store sells consists of `1.5m`. Some stores like to show the `1.5m` on the NumberInput. Also, when a user clicks the plus button, they want to show `3m`, not `2.5m`, since they measure their unit per `1.5m`.

You are doing some things right: you shouldn't vary the molecule's value based on the unit multiplier. It should always count how many units of `1.5m` are selected, that is, it should always be an integer.

Reply:
I agree with it.